### DISCUSSION TITLE
License discussion

### DISCUSSION THREAD

**Discussion by phinguyen202** [CoderLine/alphaTab#617] [General]
Hi guys,

I think we should talk about the licensing, then everyone can do things in the right way when using **alphaTab**.
**alphaTab** is licensed under the **Mozilla Public License 2.0 (MPL 2.0)**, permissions, and limitations are quite clear so we should focus on conditions including 3 following terms: 

- Disclose source
- License and copyright notice
- Same license (file)

I'm reading MPL 2.0 as well as do some research about the above terms. Two questions in my mind now:
- How do I "disclose" my modification to the world?
- I create a new file, do I need to "disclose" it?

Hope to see your comment and hope this topic is useful for everyone use **alphaTab** on their application.

Let's discuss!

**Comment by Danielku15** [comment]
Hi there. 
I fear that this is quite a complicated topic and I am also not a legal expert when it comes to handling OSS 100% right. In general I usually rather recommend people to consult some license experts and websites which try to decipher these license text for more normal humans 😅 

https://opensource.stackexchange.com/ is a good starting point where people ask questions about licenses and what this really means. 

There are also websites like the following trying to explain a bit what it means: 
https://tldrlegal.com/license/mozilla-public-license-2.0-(mpl-2)
https://choosealicense.com/licenses/mpl-2.0/

In general my intention behind having MPL for alphaTab is the following: 

1. I want to avoid that people just take alphaTab and "sell" it as their own product. I invest a lot of time and effort into this project and I make it available for free, give support and realize feature requests for the people. So I see it as the minimum of good behavior to keep the alphaTab watermark or tell people in some other way that this music notation display is powered by alphaTab. People often make money with their apps and platforms, and they don't give back anything to me. Making this bit of advertisment is the minimum. 

Normally software is legally required to list all 3rd party open source licenses they used to make a software or app. But in the web this is an often forgotten and misused practice to make a software without attributing the libraries used. In the world of minimizing+bundling also all copyright notices in the shipped assets of the app are stripped. This means usually you have no clue what libraries are used to build an app. This is why alphaTab enforces this watermark being rendered to keep this minimum of advertisement for the effort I invested into this product and the minimum of "giving back something" I expect from somebody using alphaTab. 

2. In Open Source everything is about collaboration. But many licenses do not really enforce people to give back. You can take the library, rewrite it, change it, extend it with amazing features but the original project will never benefit from this. Forks can be good, but without feeding things back to the main project this is for me often just plagiarism than a nice "working together culture". Coming to MPL it has a strong copyleft, meaning that all changes you make to the library, must again be published as open source and licensed under MPL. This at least gives me personally the chance to look at forks and port back good things to the upstream and benefit from the extensions they made to my library. 

The disclose source is something coming heavily form the official OSS definition and it's not so important to me. I think this is one of the rules nobody really follows and is hardly possible in the way software is shipped nowadays. I would say as long as the original unmodified version of alphaTab is used, the normal license notices are sufficient. The source code of the official releases are there and made available by putting a link to our GitHub repo and/or website. But if you made modifications things get again more important, because MPL requires you to publish all modifications under MPL, therefore you need to make your sources again available to everyone. 

Coming to your concrete questions:

> How do I "disclose" my modification to the world?

The best thing would be to follow the open source culture and contribute it to the original project. Make a fork on GitHub, add your changes to this fork. and create a pull request to bring it back to the main product. This has some benefit for you because I will also participate in maintaining these features (and reduce your effort) and also other will benefit from the work you've done. 

I think it is a risky mentality if you want to take an open source library, modify it, but make it yours and special only. This goes against the idea behind OSS. 

> I create a new file, do I need to "disclose" it?

This definition is a bit vague. But usually, if you are working within the library codebase, extend it, adapt it. You will need to disclose it because it is a modification to the open source library. But if you are able to take the final alphaTab version and plug your extensions on top with public extension points, things do not need to be disclosed. e.g. if you make a super fancy all-in-one control similar to the player on our website, but provide it as hosted script solution (like Soundslice) there is no need to license this component under MPL. You are then just using the library as it is.

**Reply by phinguyen202** [reply]
Many thanks for your comment.
This info is very useful for me to get an understanding of the idea behind the license. Now I think everyone can know what have to do for using alphaTab on their website legally or more legally.

I understand your position and all things required are reasonable. Open-source culture makes the world better somehow.

In a developing country like Vietnam, the license is usually ignored, even in some "professional" companies, but now everybody pays more attention to it. That is a good signal. 

Finally, thanks for your awesome work, I have to say that alphaTab is the best music notation library I've found on Github.
